lib/ASTGen/Sources/ASTGen/DiagnosticsBridge.swift

```diff
@@ -280,18 +280,33 @@ public func addQueuedDiagnostic(
       return false
     }
 
-    return rawPosition >= baseAddress && rawPosition < baseAddress + sf.buffer.count
+    return rawPosition >= baseAddress && rawPosition <= baseAddress + sf.buffer.count
   }
   guard let sourceFile = sourceFile else {
     // FIXME: Hard to report an error here...
     return
   }
 
-  // Find the token at that offset.
   let sourceFileBaseAddress = UnsafeRawPointer(sourceFile.buffer.baseAddress!)
   let sourceFileEndAddress = sourceFileBaseAddress + sourceFile.buffer.count
   let offset = rawPosition - sourceFileBaseAddress
-  guard let token = sourceFile.syntax.token(at: AbsolutePosition(utf8Offset: offset)) else {
+  let position = AbsolutePosition(utf8Offset: offset)
+
+  // Find the token at that offset.
+  let node: Syntax
+  if let token = sourceFile.syntax.token(at: position) {
+    node = Syntax(token)
+  } else if position == sourceFile.syntax.endPosition {
+    // FIXME: EOF token is not included in '.token(at: position)'
+    // We might want to include it, but want to avoid special handling.
+    // Also 'sourceFile.syntax' is not guaranteed to be 'SourceFileSyntax'.
+    if let token = sourceFile.syntax.lastToken(viewMode: .all) {
```

Comment:
I assume you want `viewMode: .sourceAccurate` here because you’d want to return the last token that actually has a source code representation, not a missing token.

Reply:
It's `.all` because, for example, when the syntax is a `CodeBlockSyntax` with missing `}`, the compiler would diagnose at the _end_ of the `CodeBlockSyntax` which is probably a `.missing` token. I don't think that would happen at this point, but after correctly implementing macro expansion buffer parsing in ASTGen, it might.

That being said, it doesn't actually matter anyway, since we specify `position`,  `node` only used to get the `root` syntax from it.

Reply:
Ah, good point. I agree that `.all` is the correct choice then.